### PR TITLE
feat: added multi-select fields to create multiple pricing rules.

### PR DESCRIPTION
**Pull request by noahjacob**
**Feature:**

-  User can select multiple options which creates multiple pricing rules.
creating

- On updating promotional scheme, creates new pricing rules if a new option is added.
adding_new

- Updates all pricing rules.
updates

- **[Docs](https://github.com/frappe/erpnext_documentation/pull/334)**

### CODE REVIEW

File: erpnext/accounts/doctype/promotional_scheme/promotional_scheme.py [hunk 1]
```diff
@@ -73,42 +73,73 @@ def get_pricing_rules(doc, rules = {}):
...
+				pr = frappe.new_doc("Pricing Rule")
+				pr.title = make_autoname("{0}/.####".format(doc.name))
+				temp_args = args.copy()
+				temp_args[applicable_for] = args[applicable_for][i]
+				pr = set_args(temp_args,pr,doc,child_doc,discount_fields,d)
+				new_doc.append(pr)
 
 	return new_doc
 
+def set_args(args,pr,doc,child_doc,discount_fields,d):
```
Comment (rohitwaghchaure) [inline erpnext/accounts/doctype/promotional_scheme/promotional_scheme.py]: Add space after comma

Comment (noahjacob) [reply erpnext/accounts/doctype/promotional_scheme/promotional_scheme.py]: fixed and also changed variable name of d 
`def set_args(args, pr, doc, child_doc, discount_fields, child_doc_fields):`

File: erpnext/accounts/doctype/promotional_scheme/promotional_scheme.py [hunk 2]
```diff
@@ -30,17 +30,17 @@ def validate(self):
 			frappe.throw(_("Price or product discount slabs are required"))
 
 	def on_update(self):
-		data = frappe.get_all('Pricing Rule', fields = ["promotional_scheme_id", "name"],
-			filters = {'promotional_scheme': self.name}) or {}
-
+		data = frappe.get_all('Pricing Rule', fields = ["promotional_scheme_id", "name", "creation"], order_by = 'creation asc',
+			filters = {'promotional_scheme': self.name, 'applicable_for': self.applicable_for}) or {}
```
Comment (marination) [inline erpnext/accounts/doctype/promotional_scheme/promotional_scheme.py]:
```suggestion
		data = frappe.get_all(
		   'Pricing Rule',
		   fields=["promotional_scheme_id", "name", "creation"],
		   filters = {
		       'promotional_scheme': self.name,
		       'applicable_for': self.applicable_for
		   },
		   order_by = 'creation asc'
	        ) or {}
```

File: erpnext/accounts/doctype/promotional_scheme/promotional_scheme.py [hunk 3]
```diff
@@ -73,42 +73,73 @@ def get_pricing_rules(doc, rules = {}):
 def _get_pricing_rules(doc, child_doc, discount_fields, rules = {}):
 	new_doc = []
 	args = get_args_for_pricing_rule(doc)
+	applicable_for = frappe.scrub(doc.get('applicable_for'))
 	for d in doc.get(child_doc):
 		if d.name in rules:
-			pr = frappe.get_doc('Pricing Rule', rules.get(d.name))
+			for applicable_for_value in args.get(applicable_for):
+				
+				temp_args = args.copy()
```
Comment (marination) [inline erpnext/accounts/doctype/promotional_scheme/promotional_scheme.py]:
```suggestion
			for applicable_for_value in args.get(applicable_for):
				temp_args = args.copy()
```

File: erpnext/accounts/doctype/promotional_scheme/promotional_scheme.py [hunk 4]
```diff
@@ -73,42 +73,73 @@ def get_pricing_rules(doc, rules = {}):
...
+				docname = frappe.get_all('Pricing Rule', fields = ["promotional_scheme_id", "name", applicable_for],
+					filters = {'promotional_scheme_id': d.name, applicable_for: applicable_for_value})
```
Comment (marination) [inline erpnext/accounts/doctype/promotional_scheme/promotional_scheme.py]: Format better

File: erpnext/accounts/doctype/promotional_scheme/promotional_scheme.py [hunk 5]
```diff
@@ -73,42 +73,73 @@ def get_pricing_rules(doc, rules = {}):
...
 	for d in pricing_rule_fields:
-		args[d] = doc.get(d)
-
+		if d == applicable_for:
+			items = []
+			for applicable_for_values in doc.get(applicable_for):
+				items.append(applicable_for_values.get(applicable_for))	
+			args[d] = items	
+		
+		else:
```
Comment (marination) [inline erpnext/accounts/doctype/promotional_scheme/promotional_scheme.py]:
```suggestion
			args[d] = items	
		else:
```

**Review by marination — CHANGES_REQUESTED** [review]

File: erpnext/accounts/doctype/campaign_item/campaign_item.py [hunk 6]
```diff
@@ -0,0 +1,10 @@
+# -*- coding: utf-8 -*-
+# Copyright (c) 2021, Frappe Technologies Pvt. Ltd. and contributors
+# For license information, please see license.txt
+
+from __future__ import unicode_literals
```
Comment (ankush) [inline erpnext/accounts/doctype/campaign_item/campaign_item.py]:
```suggestion
# Copyright (c) 2021, Frappe Technologies Pvt. Ltd. and contributors
# For license information, please see license.txt

```

File: erpnext/accounts/doctype/customer_group_item/customer_group_item.py [hunk 7]
```diff
@@ -0,0 +1,10 @@
+# -*- coding: utf-8 -*-
+# Copyright (c) 2021, Frappe Technologies Pvt. Ltd. and contributors
+# For license information, please see license.txt
+
+from __future__ import unicode_literals
```
Comment (ankush) [inline erpnext/accounts/doctype/customer_group_item/customer_group_item.py]:
```suggestion
# Copyright (c) 2021, Frappe Technologies Pvt. Ltd. and contributors
# For license information, please see license.txt

```

File: erpnext/accounts/doctype/territory_item/territory_item.py [hunk 8]
```diff
@@ -0,0 +1,10 @@
+# -*- coding: utf-8 -*-
+# Copyright (c) 2021, Frappe Technologies Pvt. Ltd. and contributors
+# For license information, please see license.txt
+
+from __future__ import unicode_literals
```
Comment (ankush) [inline erpnext/accounts/doctype/territory_item/territory_item.py]:
```suggestion
# Copyright (c) 2021, Frappe Technologies Pvt. Ltd. and contributors
# For license information, please see license.txt

```

**Review by ankush — CHANGES_REQUESTED** [review]
remove py2 code. Not required anymore.

File: erpnext/accounts/doctype/promotional_scheme/promotional_scheme.py [hunk 9]
```diff
@@ -73,42 +80,77 @@ def get_pricing_rules(doc, rules = {}):
...
+					}
+				)
+					
+				if docname:
+					pr = frappe.get_doc('Pricing Rule', docname[0].get('name'))
+					temp_args[applicable_for] = applicable_for_value
+					pr = set_args(temp_args, pr, doc, child_doc, discount_fields, d)
+				else:
+					pr = frappe.new_doc("Pricing Rule")
+					pr.title = make_autoname("{0}/.####".format(doc.name))
```
Comment (marination) [inline erpnext/accounts/doctype/promotional_scheme/promotional_scheme.py]: why are we forcing this naming as the title here ? Shouldn't it automatically pick this up from whatever is configured ?

**Comment by mohsinalimat** [comment]
Really this is good feature and I am waiting to use this feature.

File: erpnext/accounts/doctype/promotional_scheme/promotional_scheme.py [hunk 10]
```diff
@@ -73,42 +80,77 @@ def get_pricing_rules(doc, rules = {}):
...
-				pr.set(field, [])
-
-			apply_on = frappe.scrub(doc.get('apply_on'))
-			for d in doc.get(field):
-				pr.append(field, {
-					apply_on: d.get(apply_on),
-					'uom': d.uom
-				})
+		else:
+			for i in range(len(args.get(applicable_for))) :
```
Comment (marination) [inline erpnext/accounts/doctype/promotional_scheme/promotional_scheme.py]: Assuming `args.get(applicable_for)` is a list, just loop over the list directly.
```suggestion
                      applicable_for = args.get(applicable_for) or []
			for applicable_for_value in applicable_for:
```

-------------------------------
**Another Note**
if selling/buying is not selected and theres no `applicable_for` OR if `applicable_for` is just not selected as it is optional:
Screenshot 2021-08-10 at 4 36 51 PM

Also, there is a validation for selecting wither selling/buying in pricing rule, add it here too?

File: erpnext/accounts/doctype/promotional_scheme/promotional_scheme.py [hunk 11]
```diff
@@ -30,17 +30,24 @@ def validate(self):
 			frappe.throw(_("Price or product discount slabs are required"))
 
 	def on_update(self):
-		data = frappe.get_all('Pricing Rule', fields = ["promotional_scheme_id", "name"],
-			filters = {'promotional_scheme': self.name}) or {}
-
+		data = frappe.get_all(
```
Comment (marination) [inline erpnext/accounts/doctype/promotional_scheme/promotional_scheme.py]: `data` that is being passed around is making i hard to comprehend what it contains
```suggestion
		pricing_rules = frappe.get_all(
```

File: erpnext/accounts/doctype/promotional_scheme/promotional_scheme.py [hunk 12]
```diff
@@ -30,17 +30,24 @@ def validate(self):
...
+			'Pricing Rule',
+			fields = ["promotional_scheme_id", "name", "creation"],
+			filters = {
+				'promotional_scheme': self.name,
+				'applicable_for': self.applicable_for
+			},
+			order_by = 'creation asc',
+		) or {}
 		self.update_pricing_rules(data)
```
Comment (marination) [inline erpnext/accounts/doctype/promotional_scheme/promotional_scheme.py]:
```suggestion
		self.update_pricing_rules(pricing_rules)
```

File: erpnext/accounts/doctype/promotional_scheme/promotional_scheme.py [hunk 13]
```diff
@@ -30,17 +30,24 @@ def validate(self):
...
 
 	def update_pricing_rules(self, data):
```
Comment (marination) [inline erpnext/accounts/doctype/promotional_scheme/promotional_scheme.py]:
```suggestion
	def update_pricing_rules(self, pricing_rules):
```

File: erpnext/accounts/doctype/promotional_scheme/promotional_scheme.py [hunk 14]
```diff
@@ -30,17 +30,24 @@ def validate(self):
...
 		rules = {}
 		count = 0
-
+		names = []
 		for d in data:
```
Comment (marination) [inline erpnext/accounts/doctype/promotional_scheme/promotional_scheme.py]:
```suggestion
		for rule in pricing_rules:
```

File: erpnext/accounts/doctype/promotional_scheme/promotional_scheme.py [hunk 15]
```diff
@@ -30,17 +30,24 @@ def validate(self):
...
-			rules[d.get('promotional_scheme_id')] = d.get('name')
+			names.append(d.name)
+			rules[d.get('promotional_scheme_id')] = names
```
Comment (marination) [inline erpnext/accounts/doctype/promotional_scheme/promotional_scheme.py]:
```suggestion
			names.append(rule.name)
			rules[rule.get('promotional_scheme_id')] = names
```

File: erpnext/accounts/doctype/promotional_scheme/promotional_scheme.py [hunk 16]
```diff
@@ -73,42 +80,77 @@ def get_pricing_rules(doc, rules = {}):
...
-					'uom': d.uom
-				})
+		else:
+			for i in range(len(args.get(applicable_for))) :
 
-		new_doc.append(pr)
+				pr = frappe.new_doc("Pricing Rule")
+				pr.title = doc.name
+				temp_args = args.copy()
+				temp_args[applicable_for] = args[applicable_for][i]
```
Comment (marination) [inline erpnext/accounts/doctype/promotional_scheme/promotional_scheme.py]:
```suggestion
				temp_args[applicable_for] = applicable_for_value
```

**Review by marination — CHANGES_REQUESTED** [review]

File: erpnext/accounts/doctype/coupon_code/test_coupon_code.py [hunk 17]
```diff
@@ -57,7 +57,7 @@ def test_create_test_data():
 		})
 		item_price.insert()
 	# create test item pricing rule
-	if not frappe.db.exists("Pricing Rule","_Test Pricing Rule for _Test Item"):
+	if not frappe.db.exists("Pricing Rule",{"title": "_Test Pricing Rule for _Test Item"}):
```
Comment (marination) [inline erpnext/accounts/doctype/coupon_code/test_coupon_code.py]: Space after comma is the norm
```suggestion
	if not frappe.db.exists("Pricing Rule", {"title": "_Test Pricing Rule for _Test Item"}):
```

File: erpnext/accounts/doctype/coupon_code/test_coupon_code.py [hunk 18]
```diff
@@ -86,14 +86,15 @@ def test_create_test_data():
 		sales_partner.insert()
 	# create test item coupon code
 	if not frappe.db.exists("Coupon Code", "SAVE30"):
+		pricing_rule = frappe.db.get_value("Pricing Rule",{"title": "_Test Pricing Rule for _Test Item"},['name'])
```
Comment (marination) [inline erpnext/accounts/doctype/coupon_code/test_coupon_code.py]:
```suggestion
		pricing_rule = frappe.db.get_value("Pricing Rule", {"title": "_Test Pricing Rule for _Test Item"}, ['name'])
```

File: erpnext/accounts/doctype/promotional_scheme/promotional_scheme.py [hunk 19]
```diff
@@ -25,22 +25,31 @@
 
 class PromotionalScheme(Document):
 	def validate(self):
+		if not self.selling and not self.buying:
+			frappe.throw(_("Atleast one of the Selling or Buying must be selected"))
```
Comment (marination) [inline erpnext/accounts/doctype/promotional_scheme/promotional_scheme.py]:
```suggestion
			frappe.throw(_("Either 'Selling' or 'Buying' must be selected"), title=_("Mandatory"))
```

**Review by marination — APPROVED** [review]

**Pull request merged**